### PR TITLE
Add a timeout for opening an SSH channel

### LINKED ISSUE #2009: open_sftp() can block up to an hour
Hey all, I want to report a bug that @VakarisZ and me found.

It comes from `open_sftp` function on SSHClient which can block up to an hour, if a machine is killed after connecting

To reproduce it just connect to a VM machine, set breakpoint before calling `open_sftp`, pause the VM machine and run the `open_sftp`. It will wait up to an hour until it timeouts. I have waited 14 minutes and then unpaused the VM which unblocked the `open_sftp` function.

What I purpose for a solution is that we add timeout argument to `open_sftp`.

Here is a snippet code I used to reproduce the block:

```import paramiko
import datetime
import pdb

ssh = paramiko.SSHClient()
ssh.set_missing_host_key_policy(paramiko.WarningPolicy)

ssh.connect("x.x.x.x", port=22, username=username, password=password, timeout=10,
            auth_timeout=10, banner_timeout=5)

print(datetime.datetime.now())

pdb.set_trace()

with ssh.open_sftp() as ftp:
    print(datetime.datetime.now())


print(datetime.datetime.now())
```

### PR DESCRIPTION
**Pull request by VakarisZ**
Fixes #2009

### CODE REVIEW

File: paramiko/client.py [hunk 1]
```diff
@@ -311,6 +312,8 @@ def connect(
             for the SSH banner to be presented.
         :param float auth_timeout: an optional timeout (in seconds) to wait for
             an authentication response.
+        :param float channel_timeout: an optional timeout (in seconds) to wait
+             for a channel open response.
```
Comment (mssalvatore) [inline paramiko/client.py:316]: @VakarisZ It looks like the aligment of "for a channel open response." is off by one space. Compare with the above and below parameters.

File: paramiko/win_openssh.py [hunk 2]
```diff
@@ -18,23 +18,39 @@
...
+            try:
+                self._pipe = os.open(PIPE_NAME, os.O_RDWR | os.O_BINARY)
+            except OSError as e:
+                # retry when errno 22 which means that the server has not
+                # called DisconnectNamedPipe() yet.
+                if e.errno != 22:
+                    raise
+            else:
+                break
+            time.sleep(0.1)
```
Comment (mssalvatore) [inline paramiko/win_openssh.py:47]: @VakarisZ Did you mean to add these changes to this PR?

**Comment by bskinn** [comment]
@VakarisZ I've flagged this issue and PR to @bitprophet's attention -- he'll prioritize it among the other issues outstanding on paramiko, invoke and fabric in his weekly OSS time.

**Comment by mssalvatore** [comment]
@bskinn Is there anything we can do make this easier to merge? Any kind of testing you'd like us to do?

I know there's a lot of work below the surface for any PR. Can you bring any of that to the surface and offload it to us?

**Comment by bskinn** [comment]
It *would* be good to have some sort of test(s) added for this, if it's ~easy to do.

Beyond that, @mssalvatore, there really isn't much else that could be offloaded, AFAIK. But, thanks for offering! The first priority is the drop of Python 2 and old 3.x, and then I think after that the key/auth rewrite of #387. At that point, I think general non-critical feature/fix PRs will start to come back on the table.

**Comment by mssalvatore** [comment]
> It would be good to have some sort of test(s) added for this, if it's ~easy to do.

I'll see what we can do.

**Comment by mssalvatore** [comment]
@bskinn We've added a test! Special thanks to @ilija-lazoroski.

Please let us know if it's not sufficient. Is there anything else we can do to help grease the wheels?

**Comment by bskinn** [comment]
@mssalvatore Sorry for the slow response, and thanks @ilija-lazoroski for adding the test!

I don't think there's anything else you can do at this point -- it just now needs to make it to the front of bitprophet's work queue. Unfortunately, I can't say when that will be...but it'll be here when he's read for it.

**Comment by mssalvatore** [comment]
@VakarisZ Are you able to change the source branch from `VakarisZ:main` to `VakarisZ:ssh-channel-timeout`?

**Comment by mssalvatore** [comment]
@bskinn @bitprophet This has been rebased on 3.0.0

File: paramiko/client.py (hunk 1 above)
Comment (bskinn) [reply paramiko/client.py:316]:
```suggestion
        :param float channel_timeout: an optional timeout (in seconds) to wait
            for a channel open response.
```

**Comment by bitprophet** [comment]
Thanks for this! It'll be released shortly.

**Comment by mssalvatore** [comment]
Thanks, @bitprophet

**Pull request merged**